packages/fetchai/connections/p2p_libp2p/dht/dhtclient/dhtclient.go

```diff
@@ -414,65 +512,290 @@ func (dhtClient *DHTClient) RouteEnvelope(envel *aea.Envelope) error {
+	case *dhtnode.AcnMessage_Status:
+		status = pl.Status
+	default:
+		err = errors.New("Unexpected Acn Message")
+		lerror(err).
+			Str("op", "route").
+			Str("target", target).
+			Msg("while deserializing acn confirmation message")
+		return err
+	}
+
+	if status.Code != dhtnode.Status_SUCCESS {
+		err = errors.New(status.Code.String() + " : " + strings.Join(status.Msgs, ":"))
+		lerror(err).
+			Str("op", "route").
+			Str("target", target).
+			Msg("failed to deliver envelope")
+		return err
+	}
+
+	return err
 }
 
 func (dhtClient *DHTClient) handleAeaEnvelopeStream(stream network.Stream) {
 	lerror, lwarn, _, ldebug := dhtClient.getLoggers()
 
-	ldebug().Msgf("Got a new aea envelope stream")
-
-	envel, err := utils.ReadEnvelope(stream)
+	//ldebug().Msgf("Got a new aea envelope stream")
+	buf, err := utils.ReadBytes(stream)
 	if err != nil {
 		lerror(err).Msg("while reading envelope from stream")
 		err = stream.Reset()
 		ignore(err)
 		return
 	}
-	stream.Close()
+
+	// get envelope
+	msg := &dhtnode.AcnMessage{}
+	err = proto.Unmarshal(buf, msg)
+	if err != nil {
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_SERIALIZATION}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
+	}
+
+	// payload is expected to be AeaEnvelope
+	var aeaEnvelope *dhtnode.AeaEnvelope
+	switch pl := msg.Payload.(type) {
+	case *dhtnode.AcnMessage_AeaEnvelope:
+		aeaEnvelope = pl.AeaEnvelope
+	default:
+		err = errors.New("Unexpected payload")
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_UNEXPECTED_PAYLOAD}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
+	}
+
+	envel := &aea.Envelope{}
+	err = proto.Unmarshal(aeaEnvelope.Envel, envel)
+	if err != nil {
+		lerror(err).Msg("while deserializing acn aea envelope message")
+		status := &dhtnode.Status{Code: dhtnode.Status_ERROR_SERIALIZATION, Msgs: []string{err.Error()}}
+		response := &dhtnode.AcnMessage{
+			Version: dhtnode.CurrentVersion,
+			Payload: &dhtnode.AcnMessage_Status{Status: status},
+		}
+		buf, err = proto.Marshal(response)
+		ignore(err)
+		err = utils.WriteBytes(stream, buf)
+		ignore(err)
+		err = stream.Close()
+		ignore(err)
+		return
+	}
+
+	remotePubkey, err := utils.FetchAIPublicKeyFromPubKey(stream.Conn().RemotePublicKey())
+	ignore(err)
+	status, err := dhtnode.IsValidProofOfRepresentation(aeaEnvelope.Record, aeaEnvelope.Record.Address, remotePubkey)
```

Comment:
[STEP3.b] when a peer receives an incoming envelope it validate its source PoR and deliver it to local agent or return error